### PR TITLE
wg_engine: optimize with the internal triangulation algorithm and mesh rendering

### PR DESCRIPTION
**Pull request by SergeyLebedkin**
Optimizations:
- "on-the-fly" computing polyline segments lengths, total length and bboxes
- optimaze stroking algorithm by prevent vector normalizations
- tabulated sin and cos functions
- reducing system and graphics memory realocations
- prealocate triangle fans index buffers (less memory writes)
- using render meshes heaps to reduce webgpu instances re-creations
- using single instance of pilylines for standard operations such as trim and split
- "on-the-fly" strokes generations with dash pattern
- "on-the-fly" mesh objects generation in a process of path decoding
- merge strokes into single mesh by each shape

### CODE REVIEW

File: src/renderer/wg_engine/tvgWgGeometry.cpp [hunk 1]
```diff
@@ -23,126 +23,233 @@
...
-    if (positions.count <= 2) return;
-    indexes.reserve((positions.count - 2) * 3);
-    for (size_t i = 0; i < positions.count - 2; i++) {
-        indexes.push(0);
-        indexes.push(i + 1);
-        indexes.push(i + 2);
+    if (initialized) return;
+    initialized = true;
+    const uint32_t nPoints = 360;
+    sinus.reserve(360);
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgGeometry.cpp:36]: nPoints

File: src/renderer/wg_engine/tvgWgGeometry.cpp [hunk 2]
```diff
@@ -23,126 +23,233 @@
...
+};
+
+//***********************************************************************
+// WgPolyline
+//***********************************************************************
+
+WgPolyline::WgPolyline()
+{
+    pts.reserve(1024);
+    dist.reserve(1014);
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgGeometry.cpp:58]: any reason why 1014 not 1024?

File: src/renderer/wg_engine/tvgWgGeometry.cpp [hunk 3]
```diff
@@ -23,126 +23,233 @@
...
+    pmax.x = pts[imaxx].x;
+    pmax.y = pts[imaxy].y;
+}
+
+//***********************************************************************
+// WgGeometryData
+//***********************************************************************
+
+WgGeometryData::WgGeometryData() {
+    positions.pts.reserve(10240);
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgGeometry.cpp:183]: 1024? since this number appears in few places I'd use constexpr

**Review by mgrudzinska — COMMENTED** [review]
in some places there are magic numbers. would be easier to debug in the future, if we keep them in some constexpr ?
8? 3? 2? 
(sure not all of them :) )
also M_PI -> MATH_PI ?

**Comment by SergeyLebedkin** [comment]
> in some places there are magic numbers. would be easier to debug in the future, if we keep them in some constexpr ? 8? 3? 2? (sure not all of them :) ) also M_PI -> MATH_PI ?

M_PI -> MATH_PI I believe that we should use our own PI constant, you are right

**Comment by hermet** [comment]
@SergeyLebedkin Hello, could you please improve your commit messages? Even it currently does not follow our commit format.

File: src/renderer/wg_engine/tvgWgGeometry.h [hunk 4]
```diff
@@ -49,34 +49,62 @@ class WgPoint
...
+    bool initialized{};
+
+    void initialize();
+    void release();
+};
+
+
+struct WgPolyline
+{
+    // polyline points
```
Comment (hermet) [inline src/renderer/wg_engine/tvgWgGeometry.h:75]: Suggestions:
1. These comments in the WgPolyline structure merely duplicate the variable names and seem unnecessary. Comments should have any hidden purpose or intention more than it.
2. Please append comments on one line if necessary.

**Review by hermet — APPROVED** [review]

**Pull request merged**